# Change Log

## [Unreleased](https://github.com/sanger-pathogens/Bio-RNASeq/tree/HEAD)

**Merged pull requests:**

- Include Travis badge [\#61](https://github.com/sanger-pathogens/Bio-RNASeq/pull/61) ([ssjunnebo](https://github.com/ssjunnebo))
- 574328 Set up Travis CI [\#60](https://github.com/sanger-pathogens/Bio-RNASeq/pull/60) ([ssjunnebo](https://github.com/ssjunnebo))
- fix broken link [\#59](https://github.com/sanger-pathogens/Bio-RNASeq/pull/59) ([ssjunnebo](https://github.com/ssjunnebo))
- update readme format [\#58](https://github.com/sanger-pathogens/Bio-RNASeq/pull/58) ([ssjunnebo](https://github.com/ssjunnebo))
- update LICENSE [\#57](https://github.com/sanger-pathogens/Bio-RNASeq/pull/57) ([ssjunnebo](https://github.com/ssjunnebo))
- nds\_change\_to\_dist\_zilla\_starter\_bundle [\#55](https://github.com/sanger-pathogens/Bio-RNASeq/pull/55) ([nds](https://github.com/nds))
- Added ap13 and changed Jorge email to path-help [\#54](https://github.com/sanger-pathogens/Bio-RNASeq/pull/54) ([aslett1](https://github.com/aslett1))
- Dzil handling dependencies automatically [\#53](https://github.com/sanger-pathogens/Bio-RNASeq/pull/53) ([js21](https://github.com/js21))
- A couple more deps that had not been added previously [\#52](https://github.com/sanger-pathogens/Bio-RNASeq/pull/52) ([js21](https://github.com/js21))
- Added dependencies to dist.ini [\#51](https://github.com/sanger-pathogens/Bio-RNASeq/pull/51) ([js21](https://github.com/js21))
- Concatenated tabix coverage plots [\#50](https://github.com/sanger-pathogens/Bio-RNASeq/pull/50) ([js21](https://github.com/js21))
- 88% test coverage for DeSeq wrapper [\#49](https://github.com/sanger-pathogens/Bio-RNASeq/pull/49) ([js21](https://github.com/js21))
- Test mode for deseq run implemented [\#48](https://github.com/sanger-pathogens/Bio-RNASeq/pull/48) ([js21](https://github.com/js21))
- remove line if its not a header and doesnt look like read [\#47](https://github.com/sanger-pathogens/Bio-RNASeq/pull/47) ([andrewjpage](https://github.com/andrewjpage))
- strip out tags in reads [\#46](https://github.com/sanger-pathogens/Bio-RNASeq/pull/46) ([andrewjpage](https://github.com/andrewjpage))
- Optimise base positions [\#45](https://github.com/sanger-pathogens/Bio-RNASeq/pull/45) ([andrewjpage](https://github.com/andrewjpage))
- Temp dir fork data struct [\#44](https://github.com/sanger-pathogens/Bio-RNASeq/pull/44) ([andrewjpage](https://github.com/andrewjpage))
- tags to be ignored wasnt being properly constructed [\#43](https://github.com/sanger-pathogens/Bio-RNASeq/pull/43) ([andrewjpage](https://github.com/andrewjpage))
- parallel value [\#42](https://github.com/sanger-pathogens/Bio-RNASeq/pull/42) ([andrewjpage](https://github.com/andrewjpage))
- split out no read annotation [\#41](https://github.com/sanger-pathogens/Bio-RNASeq/pull/41) ([andrewjpage](https://github.com/andrewjpage))
- returned after 1st chromosome [\#40](https://github.com/sanger-pathogens/Bio-RNASeq/pull/40) ([andrewjpage](https://github.com/andrewjpage))
- Prefilter features [\#39](https://github.com/sanger-pathogens/Bio-RNASeq/pull/39) ([andrewjpage](https://github.com/andrewjpage))
- random seed after fork to avoid tmp file name clash [\#38](https://github.com/sanger-pathogens/Bio-RNASeq/pull/38) ([andrewjpage](https://github.com/andrewjpage))
- Create tmp files in cwd not /tmp [\#37](https://github.com/sanger-pathogens/Bio-RNASeq/pull/37) ([andrewjpage](https://github.com/andrewjpage))
- Parallel bams [\#36](https://github.com/sanger-pathogens/Bio-RNASeq/pull/36) ([andrewjpage](https://github.com/andrewjpage))
- Speedup rnaseq [\#35](https://github.com/sanger-pathogens/Bio-RNASeq/pull/35) ([andrewjpage](https://github.com/andrewjpage))
- Transcript is now also accepted as middle feature - Ensembl gene model [\#34](https://github.com/sanger-pathogens/Bio-RNASeq/pull/34) ([js21](https://github.com/js21))
- Added AUTHORS file [\#33](https://github.com/sanger-pathogens/Bio-RNASeq/pull/33) ([aslett1](https://github.com/aslett1))
- Support for Ensembl and new format Chado GFF files [\#32](https://github.com/sanger-pathogens/Bio-RNASeq/pull/32) ([js21](https://github.com/js21))
- Die if bam and gff are not related [\#31](https://github.com/sanger-pathogens/Bio-RNASeq/pull/31) ([js21](https://github.com/js21))
- All tests pass [\#30](https://github.com/sanger-pathogens/Bio-RNASeq/pull/30) ([js21](https://github.com/js21))
- Update README.md [\#29](https://github.com/sanger-pathogens/Bio-RNASeq/pull/29) ([js21](https://github.com/js21))
- Documentation is in. Clean up of test\_helper and genral tidy-up [\#28](https://github.com/sanger-pathogens/Bio-RNASeq/pull/28) ([js21](https://github.com/js21))
- forgot to check for exit code. [\#27](https://github.com/sanger-pathogens/Bio-RNASeq/pull/27) ([js21](https://github.com/js21))
- Now final typo on module [\#26](https://github.com/sanger-pathogens/Bio-RNASeq/pull/26) ([js21](https://github.com/js21))
- Final bsub submission [\#25](https://github.com/sanger-pathogens/Bio-RNASeq/pull/25) ([js21](https://github.com/js21))
- New DeSeq step fully implemented. Documentation will follow shortly [\#24](https://github.com/sanger-pathogens/Bio-RNASeq/pull/24) ([js21](https://github.com/js21))
- Tests needed for lack of cmd args [\#23](https://github.com/sanger-pathogens/Bio-RNASeq/pull/23) ([js21](https://github.com/js21))
- GFF3 concatenator utility added [\#22](https://github.com/sanger-pathogens/Bio-RNASeq/pull/22) ([js21](https://github.com/js21))
- Cleaned up the application [\#21](https://github.com/sanger-pathogens/Bio-RNASeq/pull/21) ([js21](https://github.com/js21))
- Last cleanup of not used modules [\#20](https://github.com/sanger-pathogens/Bio-RNASeq/pull/20) ([js21](https://github.com/js21))
- Reimplementation of the total number of reads that mapped, plus deletion of several modules [\#19](https://github.com/sanger-pathogens/Bio-RNASeq/pull/19) ([js21](https://github.com/js21))
- Added the Sam.pm module as it was actually required by BAMstats.pm [\#18](https://github.com/sanger-pathogens/Bio-RNASeq/pull/18) ([js21](https://github.com/js21))
- Minor corrections to requires section of README.md [\#17](https://github.com/sanger-pathogens/Bio-RNASeq/pull/17) ([js21](https://github.com/js21))
- Another minor formatting issue on README [\#16](https://github.com/sanger-pathogens/Bio-RNASeq/pull/16) ([js21](https://github.com/js21))
- Complete restructuring of the code [\#15](https://github.com/sanger-pathogens/Bio-RNASeq/pull/15) ([js21](https://github.com/js21))
- All tests for all protocols and total mapped read methods implemented. [\#14](https://github.com/sanger-pathogens/Bio-RNASeq/pull/14) ([js21](https://github.com/js21))
- Tests for the Expression.pm object are fully implemented. Size reduction... [\#13](https://github.com/sanger-pathogens/Bio-RNASeq/pull/13) ([js21](https://github.com/js21))
- Added new RPKM calculation method based on total reads mapped to gene mo... [\#12](https://github.com/sanger-pathogens/Bio-RNASeq/pull/12) ([js21](https://github.com/js21))
- Option handling properly implemented. Also new method for creating total\_mapped\_reads started [\#11](https://github.com/sanger-pathogens/Bio-RNASeq/pull/11) ([js21](https://github.com/js21))
- Total Mapped Reads method fucntionality added [\#10](https://github.com/sanger-pathogens/Bio-RNASeq/pull/10) ([js21](https://github.com/js21))
- Update README.md [\#9](https://github.com/sanger-pathogens/Bio-RNASeq/pull/9) ([js21](https://github.com/js21))
- Rename LegacyBranch\_README.md to README.md [\#8](https://github.com/sanger-pathogens/Bio-RNASeq/pull/8) ([js21](https://github.com/js21))
- Create LegacyBranch\_README.md [\#7](https://github.com/sanger-pathogens/Bio-RNASeq/pull/7) ([js21](https://github.com/js21))
- Redefined the module call Bio::RNASeq::StandardProtocol::Read on the Ali... [\#6](https://github.com/sanger-pathogens/Bio-RNASeq/pull/6) ([js21](https://github.com/js21))
- Can build with dzil build. All tests pass dzil test [\#5](https://github.com/sanger-pathogens/Bio-RNASeq/pull/5) ([js21](https://github.com/js21))
- Tests all pass with the new directory structure [\#4](https://github.com/sanger-pathogens/Bio-RNASeq/pull/4) ([js21](https://github.com/js21))
- Only necessary modules in repo. [\#3](https://github.com/sanger-pathogens/Bio-RNASeq/pull/3) ([js21](https://github.com/js21))
- All dependencies in place and all tests passed [\#2](https://github.com/sanger-pathogens/Bio-RNASeq/pull/2) ([js21](https://github.com/js21))
- First commit with most of the codebase in place [\#1](https://github.com/sanger-pathogens/Bio-RNASeq/pull/1) ([js21](https://github.com/js21))

